vpc/tool/container2/setup_container_linux.go

```diff
@@ -121,6 +140,241 @@ func DoSetupContainer(ctx context.Context, netnsfd int, assignment *vpcapi.Assig
+	if err == nil {
+		var fstat unix.Statfs_t
+		err = unix.Fstatfs(fd, &fstat)
+		if err != nil {
+			_ = unix.Close(fd)
+			return nil, fmt.Errorf("Unable to fstat fs: %w", err)
+		}
+		if fstat.Type == unix.NSFS_MAGIC {
+			return os.NewFile(uintptr(fd), assignment.AssignmentId), nil
+		}
+	}
+	defer unix.Close(fd)
+
+	if err != nil {
+		return nil, fmt.Errorf("Could not open assignment namespace %q: %w", assignment.AssignmentId, err)
+	}
+
+	newnetns, err := createTransitionNS(ctx, branchLink, assignment)
+	if err != nil {
+		return nil, fmt.Errorf("Could not create new transition ns: %w", err)
+	}
+	defer unix.Close(newnetns)
+
+	emptyPath, err := syscall.BytePtrFromString("")
+	if err != nil {
+		panic(err)
+	}
+
+	srctree, _, errno := syscall.Syscall(unix.SYS_OPEN_TREE, uintptr(newnetns), uintptr(unsafe.Pointer(emptyPath)), unix.AT_EMPTY_PATH|OPEN_TREE_CLONE|unix.O_CLOEXEC)
+	if errno != 0 {
+		return nil, fmt.Errorf("Could not open src tree: %s: %w", unix.ErrnoName(errno), errno)
+	}
+	defer unix.Close(int(srctree))
+
+	dsttree, _, errno := syscall.Syscall(unix.SYS_OPEN_TREE, uintptr(fd), uintptr(unsafe.Pointer(emptyPath)), unix.AT_EMPTY_PATH|unix.O_CLOEXEC)
+	if errno != 0 {
+		return nil, fmt.Errorf("Could not open dst tree: %s: %w", unix.ErrnoName(errno), errno)
+	}
+	defer unix.Close(int(dsttree))
+
+	assignmentIDPath, err := syscall.BytePtrFromString(assignment.AssignmentId)
+	if err != nil {
+		return nil, fmt.Errorf("Could not create byte string from assignment: %w", err)
+	}
+
+	_, _, errno = syscall.Syscall6(unix.SYS_MOVE_MOUNT, srctree, uintptr(unsafe.Pointer(emptyPath)), transitionNamespaceDirFile.Fd(), uintptr(unsafe.Pointer(assignmentIDPath)), MOVE_MOUNT_F_EMPTY_PATH, 0)
+	if errno != 0 {
+		return nil, fmt.Errorf("Could not move mount: %s: %w", unix.ErrnoName(errno), errno)
+	}
+
+	dup, err := unix.Dup(fd)
+	if err != nil {
+		return nil, fmt.Errorf("Cannot dup transition ns fd: %w", err)
+	}
+
+	return os.NewFile(uintptr(dup), assignment.AssignmentId), nil
+}
+
+func createTransitionNS(ctx context.Context, branchLink netlink.Link, assignment *vpcapi.AssignIPResponseV3_TransitionAssignment) (int, error) {
+	runtime.LockOSThread()
+	defer runtime.UnlockOSThread()
+
+	originalNetNS, err := netns.Get()
+	if err != nil {
+		return 0, fmt.Errorf("Could not retrieve current netns: %w", err)
+	}
+	defer originalNetNS.Close()
+
+	ns, err := netns.New()
+	if err != nil {
+		return 0, fmt.Errorf("Could not create transition ns: %w", err)
+	}
+	defer ns.Close()
+
+	handle, newHandleErr := netlink.NewHandle()
+	err = netns.Setns(originalNetNS, unix.CLONE_NEWNET)
+	if err != nil {
+		return 0, fmt.Errorf("Could not restore original network namespace: %w", err)
+	}
+
+	if newHandleErr != nil {
+		return 0, fmt.Errorf("Could not get netns handle: %w", err)
+	}
+	defer handle.Delete()
+
+	r := rand.New(rand.NewSource(time.Now().UnixNano())) // nolint: gosec
+	transitionInterfaceName := fmt.Sprintf("transition-%d", r.Intn(10000))
```

Comment:
neat!

Reply:
Kinda an optimization, but if we know the transition NS is always a unique set of SGs/subnet(?), which I guess that logic is in the vpc service itself, but it would be cool if we could give it a significant name.

Or at least you could do `transition-$assignmentID` ?

Reply:
That gets too long, interface names are limited to 14 or 15 characters.